docs/updates/5-1-2017-flow.md

```diff
@@ -0,0 +1,156 @@
+### Flow Coverage Report (May 1st)
+
+`yarn flow-coverage`
+
+Stats:
+* Flow version `0.43.1`
+* **60 %**
+
+#### Components
+
+* **54%** coverage
+* 44 components are typed (almost all)
+* `yarn flow-react` - creates a flow report for components
+
+Typing our components helps us with rendering logic, which can take advantage
+of our domain objects such as `Frame`.
+
+```js
+function renderFrameTitle(frame: Frame) {
+  const displayName = formatDisplayName(frame);
+  return dom.div({ className: "title" }, displayName);
+}
+```
+
+We are not yet typing our props and state, which would be a helpful way
+to get additional coverage. We currently use `PropTypes`, which do not include
+flow types. We will hopefully switch to [props] soon to take advantage of static typing.
+
+```js
+shouldComponentUpdate(nextProps, nextState) {
+  const { frames, selectedFrame } = this.props;
+  const { showAllFrames } = this.state;
+  return (
+    frames !== nextProps.frames ||
+    selectedFrame !== nextProps.selectedFrame ||
+    showAllFrames !== nextState.showAllFrames
+  );
+}
+```
+
+There have been a couple places where flow types have been difficult to get right.
+
+1. **contravariants** `this.toggleFramesDisplay = this.toggleFramesDisplay.bind(this)`
```

Comment:
Class methods are covariant in Flow by default, which in effect means they are "read-only" and cannot be assigned to.

```js
class Error extends React.Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this); // Error
    //   ^ property `onClick`. Covariant property `onClick` incompatible with contravariant use in
    //     assignment of property `onClick`
  }
  
  onClick(event: ButtonClickEvent) {
    // ...
  }
  
  render() {
    return <button onClick={this.onClick}>Click Me</button>;
  }
}
```

In order to make this work, you need to use an additional class field, which are invariant (read & write) by default, like this:

```js
class WorksWithClassPropertyType extends React.Component {
  constructor(props) {
    super(props);
    this.onClick = this.onClick.bind(this); // Works!
  }
  
  onClick: (ButtonClickEvent) => void;
  onClick(event) {
    // ...
  }
}
```

Alternatively you can use the class properties proposal to get around all of this:

```js
class WorksWithClassPropertyArrowFunction extends React.Component {
  onClick = (event: ButtonClickEvent) => {
    // ...
  };
}
```

[[See in Try Flow](https://flow.org/try/#0PTAEAEDMBsHsHcBQBLAtgB1gJwC6gEoCmAhgMZ6RayqgDkWJ5tA3IojgJ7qGgBCArjhywAdgGFoyUgGsAogDdCIvAF5QAWVj8AzoQVK8AMlABvRKFCl+WBsoAqxLAHNCOAFygAEnfUAZAUKistCEqAYANIgAvqyIpNDE2tqgsjbYoIQAHjhKACbJRGQ4AHRi1JgiBqbmlqLaOFj85NgAFOhU6NoAlNUWFtr83FhtHd2sfaA4ABbI2sWiElLSoGrTs-PikjLFAEbIIrkta2M1UTU1C1vSLYSKyh4BwptL+so9ZhMgoMU-p+cWtlyhGG7xqANc1hEoAAPDtBE9QJclioTMcNosZFEAHwY5bqQjQ4BwwIiLHjUBnM5xBJJUAAdWw0m0dOQ0wkiW0AAUOsDOHYuDwsjkDgVGCUyhhRFUPrURPVGs1hu1YJ1QRMBkMRiqThM0UiZCtJjM5vrpLt9odjl1yWcLBdnjIPC1HqbXjgeiosaB5LBkLlyaabnd3b0+l8fsU-lT4hz6YzmaypuyktyVbyOABBNLwABi-BE5GQogy2TyoqKpXKUuUodNhqDBge8NdwY9XplFnDvwsMWiiCAA)]